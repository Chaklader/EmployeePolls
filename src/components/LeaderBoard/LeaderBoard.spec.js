import React from "react";
import { render } from "@testing-library/react";
import { applyMiddleware, createStore, compose } from "redux";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import LeaderBoard from "./LeaderBoard";
import rootReducer from "./../../reducers";
import { handleInitialData } from "../../actions/shared";

describe("Verify that the leaderboard", () => {
  it("Displaying the correct user name, the number of questions asked, and the number of questions answered.", async () => {
    const store = createStore(rootReducer, compose(applyMiddleware(thunk)));
    await handleInitialData()(store.dispatch);
    const { users } = store.getState();
    const tree = render(
      <Provider store={store}>
        <LeaderBoard />
      </Provider>
    );

    const { container } = tree;

    const sortedUserIDs = Object.keys(users).sort((idA, idB) => {
      const scoreA =
        Object.keys(users[idA].answers).length + users[idA].questions.length;
      const scoreB =
        Object.keys(users[idB].answers).length + users[idB].questions.length;

      return scoreB - scoreA;
    });

    const found = await Promise.all(
      sortedUserIDs.map((_, index) => {
        return tree.getByTestId(`user-stat-${index}`);
      })
    );
    expect(container.childNodes[1]).toBe(found[0]);

    found.map((f, index) => {
      expect(container.childNodes[index + 1]).toBe(f);
    });

    expect(tree).toMatchSnapshot();
  }, 30000);
});
