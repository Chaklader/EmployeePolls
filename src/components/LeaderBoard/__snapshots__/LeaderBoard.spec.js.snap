// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify that the leaderboard Displaying the correct user name, the number of questions asked, and the number of questions answered. 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h2
        class="text-center my-3"
      >
        <small>
          LeaderBoard
        </small>
      </h2>
      <div
        class="justify-content-center row"
        data-testid="user-stat-0"
      >
        <div
          class="col-md-6 col-12"
        >
          <div
            class="m-3 card bg-light"
          >
            <div
              class="card-header"
            >
              <img
                alt="user avatar"
                class="mr-2 img-fluid rounded-circle"
                height="40"
                src="https://tylermcginnis.com/would-you-rather/sarah.jpg"
                width="40"
              />
              Sarah Edo
            </div>
            <div
              class="d-flex justify-content-center card-body"
            >
              <p
                class="card-text"
              >
                Answered Questions: 
                4
                <br />
                Created Questions: 
                2
              </p>
            </div>
            <div
              class="card-footer"
            >
              Score: 
              6
            </div>
          </div>
        </div>
      </div>
      <div
        class="justify-content-center row"
        data-testid="user-stat-1"
      >
        <div
          class="col-md-6 col-12"
        >
          <div
            class="m-3 card bg-light"
          >
            <div
              class="card-header"
            >
              <img
                alt="user avatar"
                class="mr-2 img-fluid rounded-circle"
                height="40"
                src="https://tylermcginnis.com/would-you-rather/dan.jpg"
                width="40"
              />
              John Doe
            </div>
            <div
              class="d-flex justify-content-center card-body"
            >
              <p
                class="card-text"
              >
                Answered Questions: 
                3
                <br />
                Created Questions: 
                2
              </p>
            </div>
            <div
              class="card-footer"
            >
              Score: 
              5
            </div>
          </div>
        </div>
      </div>
      <div
        class="justify-content-center row"
        data-testid="user-stat-2"
      >
        <div
          class="col-md-6 col-12"
        >
          <div
            class="m-3 card bg-light"
          >
            <div
              class="card-header"
            >
              <img
                alt="user avatar"
                class="mr-2 img-fluid rounded-circle"
                height="40"
                src="https://tylermcginnis.com/would-you-rather/tyler.jpg"
                width="40"
              />
              Tyler McGinnis
            </div>
            <div
              class="d-flex justify-content-center card-body"
            >
              <p
                class="card-text"
              >
                Answered Questions: 
                2
                <br />
                Created Questions: 
                2
              </p>
            </div>
            <div
              class="card-footer"
            >
              Score: 
              4
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <h2
      class="text-center my-3"
    >
      <small>
        LeaderBoard
      </small>
    </h2>
    <div
      class="justify-content-center row"
      data-testid="user-stat-0"
    >
      <div
        class="col-md-6 col-12"
      >
        <div
          class="m-3 card bg-light"
        >
          <div
            class="card-header"
          >
            <img
              alt="user avatar"
              class="mr-2 img-fluid rounded-circle"
              height="40"
              src="https://tylermcginnis.com/would-you-rather/sarah.jpg"
              width="40"
            />
            Sarah Edo
          </div>
          <div
            class="d-flex justify-content-center card-body"
          >
            <p
              class="card-text"
            >
              Answered Questions: 
              4
              <br />
              Created Questions: 
              2
            </p>
          </div>
          <div
            class="card-footer"
          >
            Score: 
            6
          </div>
        </div>
      </div>
    </div>
    <div
      class="justify-content-center row"
      data-testid="user-stat-1"
    >
      <div
        class="col-md-6 col-12"
      >
        <div
          class="m-3 card bg-light"
        >
          <div
            class="card-header"
          >
            <img
              alt="user avatar"
              class="mr-2 img-fluid rounded-circle"
              height="40"
              src="https://tylermcginnis.com/would-you-rather/dan.jpg"
              width="40"
            />
            John Doe
          </div>
          <div
            class="d-flex justify-content-center card-body"
          >
            <p
              class="card-text"
            >
              Answered Questions: 
              3
              <br />
              Created Questions: 
              2
            </p>
          </div>
          <div
            class="card-footer"
          >
            Score: 
            5
          </div>
        </div>
      </div>
    </div>
    <div
      class="justify-content-center row"
      data-testid="user-stat-2"
    >
      <div
        class="col-md-6 col-12"
      >
        <div
          class="m-3 card bg-light"
        >
          <div
            class="card-header"
          >
            <img
              alt="user avatar"
              class="mr-2 img-fluid rounded-circle"
              height="40"
              src="https://tylermcginnis.com/would-you-rather/tyler.jpg"
              width="40"
            />
            Tyler McGinnis
          </div>
          <div
            class="d-flex justify-content-center card-body"
          >
            <p
              class="card-text"
            >
              Answered Questions: 
              2
              <br />
              Created Questions: 
              2
            </p>
          </div>
          <div
            class="card-footer"
          >
            Score: 
            4
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
